pkg/registry/generic/etcd/etcd.go

```diff
@@ -440,6 +441,33 @@ func (e *Etcd) Delete(ctx api.Context, name string, options *api.DeleteOptions)
 	return e.finalizeDelete(out, true)
 }
 
+func (e *Etcd) DeleteCollection(ctx api.Context, options *api.DeleteOptions, listOptions *unversioned.ListOptions) (runtime.Object, error) {
```

Comment:
Regarding the first one - yes - it's exactly as you wrote. I think it's the cleanest semantics, although we probably need to be clear that the operation is (at least for now) NOT atomic (i.e. it can happen that some of the elements will be removed, before error is returned.

Regarding the second comment - AFAIK yes - they are removing the concept of directory.
I agree that ideally we should remove also the "directory" entry in etcd, but currently it's not doable with our API. I will extend this TODO to be clear about.
Also note, that this PR is NOT making things worse (in fact we are moving "bulk deletion" from client to server. The result in the etcd is exactly the same.